### ISSUE TITLE
CRDs missing tiered storage fields

### ISSUE THREAD

**Issue by Aransh** [redpanda-data/helm-charts#968]
### What happened?

Trying to apply a "redpanda" kind resource to the cluster, and getting these errors:

Error from server (BadRequest): error when creating "redpanda.yaml": Redpanda in version "v1alpha1" cannot be handled as a Redpanda: strict decoding error: unknown field "spec.clusterSpec.storage.tiered.config.cloud_storage_access_key", unknown field "spec.clusterSpec.storage.tiered.credentialsSecretRef"

### What did you expect to happen?

I expected the resource to be deployed while applying these 2 tiered-storage related configurations

### How can we reproduce it (as minimally and precisely as possible)?. Please include values file.

<details>

```console
apiVersion: cluster.redpanda.com/v1alpha1
kind: Redpanda
metadata:
  name: redpanda-test
spec:
  chartRef: {}
  clusterSpec:
    storage:
      tiered:
        config:
          cloud_storage_enabled: "true"
          cloud_storage_bucket: xxxx
          cloud_storage_access_key: xxxx
          cloud_storage_region: xxx
          cloud_storage_api_endpoint: xxxx
        credentialsSecretRef:
          name: xxxx
          key: xxxx
```

</details>

### Anything else we need to know?

I've verified CRDs are updated first

### Which are the affected charts?

Operator

### Chart Version(s)

<details>

```console
Operator chart version 0.4.11
```

</details>

### Cloud provider

<details>
Linode
</details>

**Comment by chrisseto** [comment]
Hey Aransh! Thanks for sharing your CRDs earlier.

It looks like the [field you're referencing](https://github.com/redpanda-data/helm-charts/pull/963/files) was only added to the Redpanda Helm chart. It's not yet made its way into the CRD. I've filed https://github.com/redpanda-data/redpanda-operator/issues/63 to track adding support for that field in the operator.

cc @RafalKorepta to fact check me.

**Comment by Aransh** [comment]
Thank you @chrisseto 
Also, keep in mind that while "credentialsSecretRef" is a new field, "cloud_storage_access_key" is not (as far as I know)

**Comment by RafalKorepta** [comment]
You both are right. I didn't add access key to the credentials secret ref and Redpanda CRD lack of the new addition.

**Comment by topatel** [comment]
any update?

**Comment by RafalKorepta** [comment]
@topatel I added missing `credentialsSecretRef` in https://github.com/redpanda-data/redpanda-operator/pull/75.

The `cloud_storage_access_key` I will open changes today in helm-chart.

**Comment by Aransh** [comment]
@RafalKorepta Just updated our CRDs and tried again with the example yaml from my issue, still same exact error... For both fields.

> Error from server (BadRequest): error when creating "Downloads/test.yml": Redpanda in version "v1alpha1" cannot be handled as a Redpanda: strict decoding error: unknown field "spec.clusterSpec.storage.tiered.config.cloud_storage_access_key", unknown field "spec.clusterSpec.storage.tiered.credentialsSecretRef"

Also note that #981 is for fetching the access key from the secretRef, which is different from the issue described here (where we are unable to pass the access key at all)

**Comment by Aransh** [comment]
I'm no expert when It comes to CRDs, but taking a look at the latest CRDs yaml:
* I see "credentialsSecretRef" was indeed added, but seems it is under "gcpSecretManager"?
* I still see no mention of "cloud_storage_access_key"

**Comment by RafalKorepta** [comment]
> I see "credentialsSecretRef" is indeed there, but seems it is under "gcpSecretManager"? am I looking at the wrong thing?

I made mistake in the definition. You are right.

> I still see no mention of "cloud_storage_access_key"

First I need to get approval for https://github.com/redpanda-data/helm-charts/pull/995. Never the less with the fix I will align the CRD to mentioned PR definition.

**Comment by RafalKorepta** [comment]
@Aransh Could you check latest release? Please update CRD and Operator version.

REF
https://github.com/redpanda-data/redpanda-operator/releases/tag/v2.1.13-23.3.4

**Comment by Aransh** [comment]
@RafalKorepta initial test didn't work, took a deeper look at the CRDs and I see the value is called "credentials_secret_ref" and not "credetialsSecretRef" as the docs show.
Now I am able to deploy the yaml, need to see if the parameters are properly applied.

**Comment by Aransh** [comment]
Seems like the operator is failing to install the chart once I add these parameters...

> Helm install failed for release redpanda/redpanda with chart redpanda@5.7.18: execution error at (redpanda/templates/statefulset.yaml:57:41): 1.073741824e+11 is invalid SI quantity Suffixes can be: b | B | k | K | m | M | g | G | Ki | Mi | Gi or without any Suffixes

Disabled it again and all looks fine, so not related to new chart version or anything else, just these parameters

**Comment by RafalKorepta** [comment]
> @RafalKorepta initial test didn't work, took a deeper look at the CRDs and I see the value is called "credentials_secret_ref" and not "credetialsSecretRef" as the docs show.

 I'm sorry. You are right. I will send fix for that.
 
 > Helm install failed for release redpanda/redpanda with chart redpanda@5.7.18: execution error at (redpanda/templates/statefulset.yaml:57:41): 1.073741824e+11 is invalid SI quantity Suffixes can be: b | B | k | K | m | M | g | G | Ki | Mi | Gi or without any Suffixes

This seems to be a problem with some configuration that represent number. I would need to see more helm values to understand this problem.

**Comment by RafalKorepta** [comment]
@Aransh Could you try with version 2.1.13-23.3.4?

https://github.com/redpanda-data/redpanda-operator/releases/tag/v2.1.14-23.3.4

**Comment by Aransh** [comment]
All is updated to latest version (operator, redpanda, CRDs), initial deployment working now, but the "credentialsSecretRef" field doesn't seem to be passed to the helmRelease... My "redpanda" yaml contains it, but it's not on my "helmRelease" yaml.

Tried a complete deletion and re-install, same issue.

I am seeing under the "post-upgrade" job logs:

> Validation errors:
cloud_storage_enabled: To enable cloud storage you need to configure S3 or Azure Blob Storage access. For S3 cloud_storage_region, cloud_storage_bucket, cloud_storage_access_key, cloud_storage_secret_key must be set. For ABS cloud_storage_azure_storage_account, cloud_storage_azure_container, cloud_storage_azure_shared_key must be set

**Comment by RafalKorepta** [comment]
> initial deployment working now, but the "credentialsSecretRef" field doesn't seem to be passed to the helmRelease... 

If operator container image is latest `2.1.13-23.3.4`, then it should just copy the values from Redpanda to helmRelease. From what I recall new fields in CRD that are not known to operator binary are omitted. 

Could you show the Redpanda snippet with `credentialsSecretRef` definition?

**Comment by Aransh** [comment]
Sure:

```
    storage:
      tiered:
        config:
          cloud_storage_api_endpoint: https://us-iad-2.linodeobjects.com
          cloud_storage_bucket: xxx-xxx
          cloud_storage_cache_size: '107374182400'
          cloud_storage_enabled: 'true'
          cloud_storage_region: us-iad-2
        credentialsSecretRef:
          accessKey:
            configurationKey: cloud_storage_access_key
            key: OBJECT_STORAGE_ACCESS_KEY
            name: redpanda-tiered-storage
          secretKey:
            configurationKey: cloud_storage_secret_key
            key: OBJECT_STORAGE_SECRET_KEY
            name: redpanda-tiered-storage
```

**Comment by RafalKorepta** [comment]
I'm sorry. https://github.com/redpanda-data/helm-charts/issues/968#issuecomment-1920690546

I wrongly communicated `2.1.13-23.3.4`.

Please verify that you are using `2.1.14-23.3.4`

## Reference

https://github.com/redpanda-data/redpanda-operator/releases/tag/v2.1.14-23.3.4

**Comment by Aransh** [comment]
I see you haven't released an updated chart for this version yet, latest operator chart version (0.4.17) still uses v2.1.13-23.3.4

Let me try and manually update the values

**Comment by Aransh** [comment]
@RafalKorepta Indeed with this version the operator passes the parameter properly to the helmRelease, but seems it is still not properly used.
Getting same error in the post-upgrade job:
> cloud_storage_enabled: To enable cloud storage you need to configure S3 or Azure Blob Storage access. For S3 cloud_storage_region, cloud_storage_bucket, cloud_storage_access_key, cloud_storage_secret_key must be set. For ABS cloud_storage_azure_storage_account, cloud_storage_azure_container, cloud_storage_azure_shared_key must be set

**Comment by chrisseto** [comment]
Hey @Aransh! I'm taking a peak now. Rafal's day is coming to an end. I think I have an idea of what's going on. We generate a slightly incomplete version of the config (No access or secret key) that is sent to the post-upgrade job which is why you're getting the error message.

Let me confirm my findings and then I'll try to find you a work around or fix.

**Comment by chrisseto** [comment]
Okay, I think I've honed in on the issue but I'll still need to poke around a bit more to confirm. I think our configuration script is setting the keys the wrong way.

If my guess is right, you should be able to work around the issue by manually setting the secret and access keys by shelling into a Redpanda pod.

```bash
rpk cluster config set cloud_storage_access_key <ACCESS KEY>
rpk cluster config set cloud_storage_secret_key <SECRET KEY>
```

After that, the upgrade job should succeed and everything should hum along as expected.
If that does work for you, I'll PR up the fix. Otherwise, I'll nail down some proof on my end and then send up a PR.

Thanks for being so patience and communicative with us! It's clear that our test infrastructure isn't quite up to snuff. There's no quick and easy win available right now but we'll nail down a long term solution soon.

**Comment by RafalKorepta** [comment]
@Aransh Thank's to @chrisseto fix Redpanda helm chart version `5.7.21` should fix your problem. 

Can you reference in the Redpanda custom resource new version?

## Reference

https://github.com/redpanda-data/helm-charts/pull/1003

https://github.com/redpanda-data/redpanda-operator/blob/f8fe7bdf176ae236bbb372e3b273c2097b6ceece/src/go/k8s/tests/e2e-v2/resource-redpanda-crud/00-create-redpanda.yaml#L6-L8

**Comment by Aransh** [comment]
@RafalKorepta @chrisseto Hey guys, sorry was away a few days, took a look now, my findings:
* First of all, the initial upgrade to 5.7.20 failed for me, and seems the operator wasn't trying to upgrade anymore, so I had to manually uninstall and let it reinstall from scratch
* This leads me to my next issue - trying to deploy a cluster from scratch with tiered storage enabled seems to be broken, brokers were failing to start, here's the log from one of them (all containers, note I censored our domain):
> ++ SERVICE_NAME=redpanda-1
++ KUBERNETES_NODE_NAME=lke141267-207573-4726f8ad0000
++ POD_ORDINAL=1
+++ expr 1 + 1
++ BROKER_INDEX=2
++ CONFIG=/etc/redpanda/redpanda.yaml
++ cp /tmp/base-config/redpanda.yaml /etc/redpanda/redpanda.yaml
++ cp /tmp/base-config/bootstrap.yaml /etc/redpanda/.bootstrap.yaml
++ LISTENER='{"address":"redpanda-1.redpanda.redpanda.svc.cluster.local.","name":"internal","port":9093}'
++ rpk redpanda config --config /etc/redpanda/redpanda.yaml set 'redpanda.advertised_kafka_api[0]' '{"address":"redpanda-1.redpanda.redpanda.svc.cluster.local.","name":"internal","port":9093}'
++ ADVERTISED_KAFKA_ADDRESSES=()
++ PREFIX_TEMPLATE=
++ ADVERTISED_KAFKA_ADDRESSES+=("{\"address\":\"${SERVICE_NAME}. xxxxxxxx\",\"name\":\"default\",\"port\":31092}")
++ PREFIX_TEMPLATE=
++ ADVERTISED_KAFKA_ADDRESSES+=("{\"address\":\"${SERVICE_NAME}. xxxxxxxx\",\"name\":\"default\",\"port\":31092}")
++ PREFIX_TEMPLATE=
++ ADVERTISED_KAFKA_ADDRESSES+=("{\"address\":\"${SERVICE_NAME}. xxxxxxxx\",\"name\":\"default\",\"port\":31092}")
++ rpk redpanda config --config /etc/redpanda/redpanda.yaml set 'redpanda.advertised_kafka_api[1]' '{"address":"xxxxxxxx","name":"default","port":31092}'
++ LISTENER='{"address":"redpanda-1.redpanda.redpanda.svc.cluster.local.","name":"internal","port":8082}'
++ rpk redpanda config --config /etc/redpanda/redpanda.yaml set 'pandaproxy.advertised_pandaproxy_api[0]' '{"address":"redpanda-1.redpanda.redpanda.svc.cluster.local.","name":"internal","port":8082}'
++ ADVERTISED_HTTP_ADDRESSES=()
++ PREFIX_TEMPLATE=
++ ADVERTISED_HTTP_ADDRESSES+=("{\"address\":\"${SERVICE_NAME}. xxxxxxxx\",\"name\":\"default\",\"port\":30082}")
++ PREFIX_TEMPLATE=
++ ADVERTISED_HTTP_ADDRESSES+=("{\"address\":\"${SERVICE_NAME}. xxxxxxxx\",\"name\":\"default\",\"port\":30082}")
++ PREFIX_TEMPLATE=
++ ADVERTISED_HTTP_ADDRESSES+=("{\"address\":\"${SERVICE_NAME}. xxxxxxxx\",\"name\":\"default\",\"port\":30082}")
++ rpk redpanda config --config /etc/redpanda/redpanda.yaml set 'pandaproxy.advertised_pandaproxy_api[1]' '{"address":"xxxxxxxx","name":"default","port":30082}'
++ set +x
Setting cloud_storage_secret_key configuration
unable to query config schema: Get "https://redpanda-2.redpanda.redpanda.svc.cluster.local.:9644/v1/cluster_config/schema": dial tcp 10.2.0.22:9644: connect: connection refused
Error from server (BadRequest): container "redpanda" in pod "redpanda-1" is waiting to start: PodInitializing

All 3 pods have the same final error (each of them trying to connect to a different broker).
This is not a blocker, but something you might want to keep in mind...
* After finding this, I uninstalled again, deployed with tiered storage disabled, and then enabled tiered storage again, giving me the same original error I sent you:
> PROPERTY                           PRIOR                               NEW
cloud_storage_api_endpoint         <nil>                               https://us-iad-2.linodeobjects.com
cloud_storage_bucket               <nil>                               xxxxxx
cloud_storage_cache_size           0                                   107374182400
cloud_storage_enable_remote_write  false                               true
cloud_storage_enable_remote_read   false                               true
cloud_storage_enabled              false                               true
cloud_storage_region               <nil>                               us-iad-2
cloud_storage_access_key           xxxxxx
cloud_storage_secret_key           [secret]
superusers                         [xxxxxx]
Validation errors:
 \* cloud_storage_enabled: To enable cloud storage you need to configure S3 or Azure Blob Storage access. For S3 cloud_storage_region, cloud_storage_bucket, cloud_storage_access_key, cloud_storage_secret_key must be set. For ABS cloud_storage_azure_storage_account, cloud_storage_azure_container, cloud_storage_azure_shared_key must be set
No changes were made

Only difference is this time, you can see the access/secret keys were indeed applied, so that's good, but still seems to fail validation.

**Comment by chrisseto** [comment]
Thanks for all the logs @Aransh!

So I see there are two discrete issues here.

1. Creating a cluster with tiered_storage enabled by default doesn't work.
2. Enabling tiered_storage _still_ doesn't work due to a validation issue.

WRT the latter, could indicate which Pod that message is from? It looks like it would be from a `Job` Pod but I just want to confirm.

**Comment by Aransh** [comment]
@chrisseto Yes this is from the "post-upgrade" job logs

**Comment by chrisseto** [comment]
Issue 1 is a catch 22. The cluster needs to be running in order for `rpk cluster config set` to work as expected but `rpk cluster config set` needs to work for the cluster to be up.

Issue 2 is a validation quirk in `rpk`/`redpanda`.

The solution for both is the same, we'll have to move the config updates to the post install/upgrade job and submit the configuration as a batch rather than one by one.

We'll try to get a fix and some regression tests in before the end of the week. CI is being a bit fussy as of late, which has thrown a wrench into things.

For now, running `rpk cluster config` manually is going to be the most sure-fire solution if you need to get a cluster up and running.

**Comment by chrisseto** [comment]
https://github.com/redpanda-data/helm-charts/pull/1018 should unblock you @Aransh !  As mentioned, CI is being a bit fussy. The logs indicate that only a specific test is failing and otherwise installations and upgrades are working as expected.

You're welcome to tryout the PR, if convenient. Otherwise we'll get CI sorted (or verify the problem and force merge to unblock you).

**Comment by Aransh** [comment]
@chrisseto I'm using the redpanda operator, not sure how to use this PRs chart with that...
However I tried running the `rpk cluster config` commands as you suggested and then enabling tiered-storage, same issue as before, so not sure what I'm missing... upgrade job gives this log again:

> PROPERTY                           PRIOR                               NEW
cloud_storage_enable_remote_read   false                               true
cloud_storage_enable_remote_write  false                               true
cloud_storage_bucket               <nil>                               xxxx
cloud_storage_enabled              false                               true
cloud_storage_api_endpoint         <nil>                               https://us-iad-2.linodeobjects.com
cloud_storage_region               <nil>                               us-iad-2
cloud_storage_cache_size           0                                   107374182400
superusers                         [xxxx]
cloud_storage_secret_key           [secret]
cloud_storage_access_key           xxxx
Validation errors:
 \* cloud_storage_enabled: To enable cloud storage you need to configure S3 or Azure Blob Storage access. For S3 cloud_storage_region, cloud_storage_bucket, cloud_storage_access_key, cloud_storage_secret_key must be set. For ABS cloud_storage_azure_storage_account, cloud_storage_azure_container, cloud_storage_azure_shared_key must be set
No changes were made

**Comment by chrisseto** [comment]
@Aransh can you try again with the latest release: https://github.com/redpanda-data/helm-charts/releases/tag/redpanda-5.7.32?

The changes in https://github.com/redpanda-data/helm-charts/pull/1054 should have resolved your woes 🎉

**Comment by Aransh** [comment]
With the fixes in 5.7.32, and the redpanda team spotting our other issue (api endpoint must not include the protocol), this is all working now 🙌🏼
Closing, thank you
